### PR TITLE
feat: inital version of httpreceiver

### LINKED ISSUE #192: Add https receiver.

### PR DESCRIPTION
**Pull request by nityanandagohain**
fixes #192 

Two implementations are present.

1) For heroku :- [Heroku ](https://devcenter.heroku.com/articles/log-drains#https-drains), here first we get the log lines by using otect counting method on the payload and then parse it using regex.
2) Any http payload :- Here we try to split the payload by newline and extract log lines
3) google :- not implementing now

### CODE REVIEW

**Comment by ankitnayan** [comment]
Is the basic auth solved?

**Comment by nityanandagohain** [comment]
That will be done at the edge level. @ankitnayan  can you share me the code where you are doing current auth using access token ?

**Comment by ankitnayan** [comment]
> That will be done at the edge level. @ankitnayan can you share me the code where you are doing current auth using access token ?

access token works directly at query-service level. Is this something you are looking at?

**Comment by nityanandagohain** [comment]
I think yes. To be precise like where do we check if the ingestion token is correct and allow ingestion from client ?

**Comment by ankitnayan** [comment]
> I think yes. To be precise like where do we check if the ingestion token is correct and allow ingestion from client ?

For access-token it happens at the query-service. You just need to pass `SIGNOZ-API-KEY` key with value in the header. 
The middleware that handles this is at https://github.com/SigNoz/signoz/blob/develop/ee/query-service/app/server.go#L301-L318. I have tested this now.

**Comment by nityanandagohain** [comment]
The auth mechanism for this is tested on staging https://github.com/SigNoz/saas/pull/21. This is ready for review.

**Review by prashant-shahi — APPROVED** [review]

File: receiver/httpreceiver/bodyparser/default.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+	ld := plog.NewLogs()
+	data := string(body)
+	if data == "" {
+		return ld, 0
+	}
+	rl := ld.ResourceLogs().AppendEmpty()
+	sl := rl.ScopeLogs().AppendEmpty()
+	loglines := strings.Split(data, "\n")
+	for _, log := range loglines {
+		sl.LogRecords().AppendEmpty().Body().SetStr(log)
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply receiver/httpreceiver/bodyparser/default.go]: Yep added to both

File: receiver/httpreceiver/httpsreceiver.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+	receiverLock.Unlock()
+
+	r.RegisterLogsConsumer(consumer)
+
+	return r, nil
+}
+
+var receiverLock sync.Mutex
+
+var receivers = map[*Config]*httpreceiver{}
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply receiver/httpreceiver/httpsreceiver.go:72]: Yeah its to ensure that same receiver is not created again, but I am not sure if that will happen. ( took this from the signalfx receiver)

File: receiver/httpreceiver/httpsreceiver_test.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+func TestCreateReceiverLogsFirst(t *testing.T) {
+	factory := NewFactory()
+	cfg := factory.CreateDefaultConfig().(*Config)
+	cfg.Endpoint = "localhost:1" // Endpoint is required, not going to be used here.
+
+	lReceiver, err := factory.CreateLogsReceiver(context.Background(), receivertest.NewNopCreateSettings(), cfg, consumertest.NewNop())
+	assert.Nil(t, err, "receiver creation failed")
+	assert.NotNil(t, lReceiver, "receiver creation failed")
+}
```
[the remark this answers is not in the document]
Comment (nityanandagohain) [reply receiver/httpreceiver/httpsreceiver_test.go:28]: Yeah let me create an issue.

**Comment by srikanthccv** [comment]
I would rename this to httplogreceiver or something. The current name confuses the http receiver used in health check monitoring.

**Comment by nityanandagohain** [comment]
Are you talking about this ? https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/httpcheckreceiver

**Comment by srikanthccv** [comment]
Yes

**Comment by nityanandagohain** [comment]
But one is `httpcheckreceiver` and ours is `httpreceiver` does it really confuse ? Or you are suggesting that `httpreceiver` is a generic name so we should go with `httplogreceiver` ?

**Comment by srikanthccv** [comment]
> does it really confuse ?

Yes, it does. I have seen people use `httpcheckreceiver` and `httpreceiver` interchangeably. 

> you are suggesting that httpreceiver is a generic name so we should go with httplogreceiver

Yes, the name should be explicit that it is an http log receiver. Had it supported other signals it could have been fine, but given there is always some protocol (jaeger/zipkin/statsd) for metrics/spans, I don't think we will ever add support for other signals to this receiver so it will remain http log receiver.

**Pull request merged**